File: In-memory-database/Level 3/C#/record.cs
// Record.cs
using System.Collections.Generic;
using System.Linq;

public class Record
{
    private readonly Dictionary<string, RecordField> _fields = new();

    public void SetField(string field, string value, int? timestamp = null, int? ttl = null)
    {
        int? expiresAt = ttl.HasValue && timestamp.HasValue ? timestamp + ttl : null;
        _fields[field] = new RecordField(value, expiresAt);
    }

    public string GetField(string field, int? timestamp = null)
    {
        if (!_fields.ContainsKey(field)) return "";
        var entry = _fields[field];
        return (timestamp == null || entry.IsValid(timestamp.Value)) ? entry.Value : "";
    }

    public bool DeleteField(string field, int? timestamp = null)
    {
        if (!_fields.ContainsKey(field)) return false;
        if (timestamp != null && !_fields[field].IsValid(timestamp.Value)) return false;
        return _fields.Remove(field);
    }

    public string Scan(int? timestamp = null)
    {
        return FormatFields(_fields
            .Where(kv => timestamp == null || kv.Value.IsValid(timestamp.Value))
            .ToDictionary(kv => kv.Key, kv => kv.Value));
    }

    public string ScanByPrefix(string prefix, int? timestamp = null)
    {
        var filtered = _fields
            .Where(kv => kv.Key.StartsWith(prefix)
                      && (timestamp == null || kv.Value.IsValid(timestamp.Value)))
            .ToDictionary(kv => kv.Key, kv => kv.Value);
        return FormatFields(filtered);
    }

    private string FormatFields(Dictionary<string, RecordField> fields)
    {
        return string.Join(", ",
            fields.OrderBy(kv => kv.Key)
                  .Select(kv => $"{kv.Key}({kv.Value.Value})"));
    }
}

File: In-memory-database/Level 3/Python/record.py
# record.py

from record_field import RecordField

class Record:
    def __init__(self):
        self.fields = {}  # field -> RecordField

    def set_field(self, field: str, value: str, timestamp: int | None = None, ttl: int | None = None):
        expires_at = timestamp + ttl if ttl is not None else None
        self.fields[field] = RecordField(value, expires_at)

    def get_field(self, field: str, timestamp: int | None = None) -> str:
        if field not in self.fields:
            return ""
        field_obj = self.fields[field]
        if timestamp is None or field_obj.is_valid(timestamp):
            return field_obj.value
        return ""

    def delete_field(self, field: str, timestamp: int | None = None) -> bool:
        if field not in self.fields:
            return False
        if timestamp is not None and not self.fields[field].is_valid(timestamp):
            return False
        del self.fields[field]
        return True

    def scan(self, timestamp: int | None = None) -> str:
        return self._format_fields(
            {f: rf for f, rf in self.fields.items()
             if timestamp is None or rf.is_valid(timestamp)}
        )

    def scan_by_prefix(self, prefix: str, timestamp: int | None = None) -> str:
        filtered = {
            f: rf for f, rf in self.fields.items()
            if f.startswith(prefix) and (timestamp is None or rf.is_valid(timestamp))
        }
        return self._format_fields(filtered)

    def _format_fields(self, fields: dict[str, RecordField]) -> str:
        return ", ".join(f"{field}({fields[field].value})" for field in sorted(fields))
